enzyme/Enzyme/MLIR/Implementations/ArithAutoDiffOpInterfaceImpl.cpp

```diff
@@ -74,11 +79,83 @@ struct AddFOpInterface
     return success();
   }
 };
-} // namespace
 
-void mlir::enzyme::registerArithDialectAutoDiffInterface(
-    DialectRegistry &registry) {
+Value addToGradient(Value oldGradient, Value addedGradient, OpBuilder & builder, MGradientUtilsReverse *gutils){
+  Value gradient = addedGradient;
+  if(gutils->hasInvertPointer(oldGradient)){
+    Value operandGradient = gutils->invertPointerM(oldGradient, builder);
+    gradient = builder.create<arith::AddFOp>(oldGradient.getLoc(), operandGradient, addedGradient);
+  }
+  gutils->mapInvertPointer(oldGradient, gradient, builder);
+}
+
+void defaultClearGradient(Operation *op, SmallVector<OpBuilder *>builders, MGradientUtilsReverse *gutils){
+  Value result = op->getOpResult(0);
+  if (gutils->invertedPointersGlobal.contains(result)){
+    FloatType floatType = result.getType().cast<FloatType>();
+    APFloat apf(floatType.getFloatSemantics(), 0);
+
+    for (OpBuilder * b : builders){
+      Value gradient = b->create<arith::ConstantFloatOp>(op->getLoc(), apf, floatType);
+      gutils->mapInvertPointer(result, gradient, *b);
+    }
+  }
+}
+
+struct AddFOpInterfaceReverse : public AutoDiffOpInterfaceReverse::ExternalModel<AddFOpInterfaceReverse, arith::AddFOp> {
+  LogicalResult createReverseModeAdjoint(Operation *op, OpBuilder &builder, MGradientUtilsReverse *gutils) const {
+    // Derivative of r = a + b -> dr = da + db
+    auto addOp = cast<arith::AddFOp>(op);
+
+    if(gutils->hasInvertPointer(addOp)){
+      Value addedGradient = gutils->invertPointerM(addOp, builder);
+      for (int i = 0; i < 2; i++) {
+        addToGradient(addOp.getOperand(i), addedGradient, builder, gutils);
+      }
```

Comment:
Nit:

```
addToGradient(addOp.getLhs(), addedGradient, builder, gutils);
addToGradient(addOp.getRhs(), addedGradient, builder, gutils);
```

both reads better and is shorter. Let's not do the magic-number operand indexing in MLIR. We do that on LLVM IR because it doesn't have better accessors.